cpp/src/arrow/compute/row/grouper.h

```diff
@@ -97,7 +97,12 @@ class ARROW_EXPORT RowSegmenter {
   virtual Status Reset() = 0;
 
   /// \brief Get the next segment for the given batch starting from the given offset
+  /// DEPRECATED: Due to its inefficiency, use GetSegments instead.
+  ARROW_DEPRECATED("Deprecated in 18.0.0. Use GetSegments instead.")
```

Comment:
Deprecate instead of removing it because I assume this is a public API (due to `grouper.h` is a public header).